controlled_access_terms.tokens.inc

```diff
@@ -0,0 +1,374 @@
+      [
+        '@date' => @$created_date->getValue($field_name)[0]['value'],
+        '@token_value' => $original,
+      ]
+    );
+  }
+  return $retval;
+}
+
+/**
+ * Gets Original File PDF file URL.
+ *
+ * @param object $node
+ *   A core drupal node object.
+ * @param string $mime_type
+ *   The name of the node's field to check for the specific relationship.
+ *
+ * @return string
+ *   The tokenized value for the given data.
+ */
+function _url_to_service_file_media_by_mimetype($node, $mime_type) {
+  $islandora_utils = \Drupal::service('islandora.utils');
+  $origfile_term = $islandora_utils->getTermForUri('http://pcdm.org/use#OriginalFile');
+  $origfile_media = $islandora_utils->getMediaWithTerm($node, $origfile_term);
+  // Get the media file's mime_type value.
+  if (is_object($origfile_media)) {
+    $origfile_mime_type = ($origfile_media->hasField('field_mime_type')) ?
+      $origfile_media->get('field_mime_type')->getValue() : NULL;
+    $origfile_mime_type = (is_array($origfile_mime_type) &&
+      array_key_exists(0, $origfile_mime_type) &&
+      is_array($origfile_mime_type[0]) &&
+      array_key_exists('value', $origfile_mime_type[0])) ?
+      $origfile_mime_type[0]['value'] : '';
+    // Compare the media file's mime_type to the given value.
+    if ($origfile_mime_type == $mime_type) {
+      $vid = $origfile_media->id();
+      if (!is_null($vid)) {
+        $media = Media::load($vid);
+        $bundle = $media->bundle();
+        // Since this is Islandora and we assume the Original File is a
+        // Document type... but doing it dynamically.
+        $fid = $media->get('field_media_' . $bundle)->getValue();
+        $fid_value = (is_array($fid) && array_key_exists(0, $fid) &&
+          array_key_exists('target_id', $fid[0])) ?
+          $fid[0]['target_id'] : NULL;
+        if (!is_null($fid_value)) {
+          $file = File::load($fid_value);
+          $url = $file->getFileUri();
+          return $url;
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Conditionally removes any commas and reverse the order of the string values.
+ *
+ * @param string $input
+ *   The string value to potentially remove and flip parts.
+ * @param bool $remove_comma
+ *   Whether or not to remove the comma and flip the parts.
+ *
+ * @return string
+ *   The value of the label.
+ */
+function controlled_access_terms_remove_comma($input, $remove_comma = FALSE) {
+  if ($remove_comma) {
+    $parts = explode(",", $input);
+    $parts = array_reverse($parts);
+    return implode(" ", $parts);
+  }
+  else {
+    return $input;
+  }
+}
+
+/**
+ * Get File referenced by given Media.
+ *
+ * @todo: Add to IslandoraUtils?
+ * @todo: Had to remove typehint MediaInterface from function.
+ *
+ * @param \Drupal\media\MediaInterface $media
+ *   The Media to operate on.
+ *
+ * @return mixed
+ *   \Drupal\file\FileInterface on success, NULL otherwise.
+ */
+function controlled_access_terms_image_from_media($media) {
```

Comment:
We can't rely on Islandora services because this module may be used independently of it.

I would crib from Islandora's Media Source Service for a Drupal solution that supports multiple Media Types (https://github.com/Islandora/islandora/blob/8.x-1.x/src/MediaSource/MediaSourceService.php#L121-L133).

Reply:
Or maybe move the chunks for getting files from Media to `islandora`?  Hacking out the particular files seems very Islandora specific.  Presumably if you were running controlled_access_terms without Islandora, you wouldn't need to jump through the same hoops to get your pdf or thumbnail.